### PR TITLE
Add option to use AWS secret in auth header for Replayer requests

### PR DESCRIPTION
**Pull request by lewijacn**
### Description
This change adds two arguments to the Replayer to add support for adding an auth header from an AWS SecretsManager secret and users `--aws-auth-header-user admin --aws-auth-header-secret arn:123`

In the near future it is expected this structure will change as the Replayer is enhanced to support different users and providing a single user auth probably won't make sense. In the meantime this gives a structure to use a secret and construct an auth header with it for replayed requests. Another detail that presented itself is that the Jolt transformation library doesn't seem to easily support dynamic value transformations (i.e. we would want to access our SecretCache regularly and have it occasionally make requests to SecretsManager to get updated secrets). For now we are only fetching the secret once at app startup but this is a temporary measure until we have support for this.

More significant changes went into the IaC than originally estimated with trying to support a range of different auth-header options, and making it a bit easier from the Copilot side to override these options with cases where we didn't setup a target cluster initially (and don't want to make a CDK change) or we just want to use a different user altogether. These changes start to step in the territory of an orchestration layer which made me a little hesitant to add, but would rather strip them out later than to be in the case of wanting to change this and not easily being able to for the short term.

### Issues Resolved
https://opensearch.atlassian.net/browse/MIGRATIONS-1191

### Testing
Manual/Unit tests

### Check List
- [x] New functionality includes testing
  - [x] All tests pass, including unit test, integration test and doctest
- [x] New functionality has been documented
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: deployment/copilot/traffic-replayer/addons/taskRole.yml [hunk 1]
```diff
@@ -47,6 +47,11 @@ Resources:
               # Values to join
               - - { "Fn::Join" : [ ":group", { "Fn::Split": [":cluster", {"Fn::ImportValue" : !Sub "${App}-${Env}-msk-cluster-arn"}]}] }
                 - "/*"
+          - Action:
+              - secretsmanager:GetSecretValue
+              - secretsmanager:DescribeSecret
+            Effect: Allow
+            Resource: "*"
```
[the remark this answers is not in the document]
Comment (lewijacn) [reply deployment/copilot/traffic-replayer/addons/taskRole.yml:54]: From my own testing/research it doesn't seem possible to specify environment variables in CFN yaml. What is normally recommended is to have a parameter in CFN that you reference when doing the CFN deploy command you would could insert an environment to override a parameter, but on our end Copilot is handling the deploy behind the scenes. Ideally where I think I wanna go is to have this value stored in SSM Parameter store and then I could reference SSM inline. Past this we may want to use SSM everywhere to replace our environment variable usage.

Please correct me if you know a better way though.

File: deployment/cdk/opensearch-service-migration/lib/opensearch-service-domain-cdk-stack.ts [hunk 2]
```diff
@@ -88,7 +93,7 @@ export class OpensearchServiceDomainCdkStack extends Stack {
       fineGrainedAccessControl: {
         masterUserArn: props.fineGrainedManagerUserARN,
         masterUserName: adminUserName,
-        masterUserPassword: adminUserSecret
+        masterUserPassword: adminUserSecret ? adminUserSecret.secretValue : undefined
```
[the remark this answers is not in the document]
Comment (lewijacn) [reply deployment/cdk/opensearch-service-migration/lib/opensearch-service-domain-cdk-stack.ts:96]: Have adjusted this to call out when this is the case. Unfortunately there doesn't seem to be a way to easily get the secret arn that will get auto-created when one is not provided. I can get the SecretValue from the Domain construct but this doesn't seem to link back to an actual Secret. I'm okay with this for now but would like to find a way around this in the future.

File: deployment/cdk/opensearch-service-migration/lib/opensearch-service-domain-cdk-stack.ts [hunk 3]
```diff
@@ -67,7 +68,11 @@ export class OpensearchServiceDomainCdkStack extends Stack {
     // Enable demo mode setting
     if (props.enableDemoAdmin) {
       adminUserName = "admin"
-      adminUserSecret = SecretValue.unsafePlainText("Admin123!")
+      adminUserSecret = new Secret(this, "demoUserSecret", {
+        secretName: "demo-user-secret",
+        // This is unsafe and strictly for ease of use in a demo mode setup
+        secretStringValue: SecretValue.unsafePlainText("Admin123!")
+      })
```
[the remark this answers is not in the document]
Comment (mikaylathompson) [reply deployment/cdk/opensearch-service-migration/lib/opensearch-service-domain-cdk-stack.ts:75]: Cool, sounds good. Maybe we can play with that down the line, but I don't think it's a concern or priority now.

File: deployment/copilot/traffic-replayer/addons/taskRole.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (mikaylathompson) [reply deployment/copilot/traffic-replayer/addons/taskRole.yml:54]: Ahh, I wasn't sure whether the env vars were evaluated before it was actually deployed.

But I like the SSM idea! I think that also opens up some convenient doors in terms of how we can redeploy or modify CDK-generated values. Maybe make a ticket for that and we'll return to it?

**Review by gregschohn — COMMENTED** [review]
I have a couple major comments/concerns.  1) Why are you initializing a secret at all?  If we don't initialize one it can't be insecure.  Code scanners might be employed to find credentials like this and flag the repo.  Worse, some users might setup an admin password to match what you're setting it to and could leave it like that (unlikely for an org that would care, but conceivable and a convenient attack for many others).  It seems fair at this point to have the user set the secret out-of-band from the any of this IAC.

Secondly, can you update this client to use AWS 2.x?  See these resources for more details
* https://mvnrepository.com/artifact/software.amazon.awssdk/secretsmanager/2.20.127
* https://github.com/awsdocs/aws-doc-sdk-examples/tree/main/javav2/example_code/secretsmanager#readme
* https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/services/secretsmanager/SecretsManagerClient.html#getSecretValue(software.amazon.awssdk.services.secretsmanager.model.GetSecretValueRequest)

Notice that links that discuss caching are only relevant for 1.x.  If you want to put caching into play, guava or a simple couple lines would probably work fine.  My take would be that refreshing isn't relevant.  If a password needs to be updated, we'll want, likely need, to rerun an entire replay.  As it is, with the cache and credentials being opaque to me, I'm concerned that you've got some synchronous calls that won't be able to be called from much of our codebase.  The 2.x client embraces netty and futures a lot more.  It also will be more consistent with the libraries in use by the rest of the codebase (like for sigv4 and hopefully for the netty code in the rest of the app in general).

**Pull request merged**